### PR TITLE
Improved ewald with cell list

### PR DESCRIPTION
**Pull request by sfogerty**
Improvements from Rene's work in Julich code implemented. 
LinkedCellList now improves real-sace energy computation.
Force calcs are improved in Ewald Sum.
Tuning changes implemented in Ewald and SPME.

(This represents more of an overhaul than we would like for one PR, but some changes were fairly coupled, and the Julich & Cabana versions had gone on a while without syncing up. Going forward, the changes should be much more incremental.)

### CODE REVIEW

**Comment by sfogerty** [comment]
I labeled this "WIP" because from what I saw, the Pthreads backend builds failed in Travis, and I still haven't worked out why.

**Comment by dalg24** [comment]
Add to whitelist

File: core/example/longrange/ewald.cpp [hunk 1]
```diff
@@ -103,163 +66,404 @@ double TEwald::compute( ParticleList &particles, double lx, double ly,
...
+                  << std::endl;
+    }
+    else
+    {
+
+        // In this branch a linked-cell list is used, which is using the Cabana
+        // implementation of linked cell neighbor list. To compute the
+        // real-space contribution the particles are sorted into cells of a
+        // given size and then for each cell the distances to particles in
+        // surrounding cells are computed. If the distance is smaller than the
```
Comment (sslattery) [inline core/example/longrange/ewald.cpp]: It seems the `Cabana::VerletList` could be used here instead. You simply set the neighborhood radius to be the cut-off distance and you will immediately get a list of all the particles with which you would compute interactions.

Comment (sslattery) [reply core/example/longrange/ewald.cpp]: yeah you could delete most of the code below and just do the force computation

**Comment by sfogerty** [comment]
@sslattery this is now ready for review

File: core/example/longrange/ewaldsum_example.cpp [hunk 2]
```diff
@@ -9,68 +9,189 @@
...
+                                                          6 );
+        domain_width( 0 ) = loc_coords.at( 0 ) * domain_size( 0 );
+        domain_width( 1 ) = ( loc_coords.at( 0 ) + 1 ) * domain_size( 0 );
+        domain_width( 2 ) = loc_coords.at( 1 ) * domain_size( 1 );
+        domain_width( 3 ) = ( loc_coords.at( 1 ) + 1 ) * domain_size( 1 );
+        domain_width( 4 ) = loc_coords.at( 2 ) * domain_size( 2 );
+        domain_width( 5 ) = ( loc_coords.at( 2 ) + 1 ) * domain_size( 2 );
+
+        // Create an empty list of all the particles
+        ParticleList *particles = new ParticleList( "particles", 10 );
```
Comment (sfogerty) [inline core/example/longrange/ewaldsum_example.cpp]: What is the 10 here?

File: core/example/longrange/ewald.cpp [hunk 3]
```diff
@@ -103,174 +102,613 @@ double TEwald::compute( ParticleList &particles, double lx, double ly,
...
 
-    auto init_p = KOKKOS_LAMBDA( const int idx ) { p( idx ) = 0.0; };
+    auto init_parameters = KOKKOS_LAMBDA( const int idx )
+    {
+        p( idx ) = 0.0;
+        f( idx, 0 ) = 0.0;
+        f( idx, 1 ) = 0.0;
+        f( idx, 2 ) = 0.0;
+    };
     Kokkos::parallel_for( Kokkos::RangePolicy<ExecutionSpace>( 0, n_max ),
```
Comment (sfogerty) [inline core/example/longrange/ewald.cpp]: Should do a deep copy on slice here

**Comment by sfogerty** [comment]
@sslattery Can this be merged?

**Pull request merged**